### PR TITLE
Type pool

### PR DESCRIPTION
**Pull request by dstallenberg**
This PR does the following:
- change the encoding structure from constructor -> methods to method -> constructor
- create a statement pool per encoding such that constants, return values, and constructor instances can be reused
- create a type pool that is used by the sampler to sample constructor instances over mocked objects
- move some of the functionality of the sampler into generator classes

### CODE REVIEW

File: libraries/analysis-javascript/lib/target/TargetVisitor.ts [hunk 1]
```diff
@@ -333,77 +333,78 @@ export class TargetVisitor extends AbstractSyntaxTreeVisitor {
...
+    //       throw new Error("unknown class method parent");
+    //     }
+    //   } else {
+    //     // unsupported
+    //     throw new Error("unknown class method parent");
+    //   }
+    // } else {
+    //   // unsupported
+    //   throw new Error("unknown class method parent");
+    // }
```
Comment (mitchellolsthoorn) [inline libraries/analysis-javascript/lib/target/TargetVisitor.ts:407]: Can you remove the commented code

File: tools/javascript/lib/commands/test.ts [hunk 2]
```diff
@@ -57,7 +58,27 @@ export function getTestCommand(
...
       "The probability we use a random type regardless of the inferred type.",
-    group: commandGroup,
+    group: typeInferenceGroup,
+    hidden: false,
+    type: "number",
+  });
+
+  options.set("reuse-statement-probability", {
+    alias: [],
+    default: 0.8,
```
Comment (mitchellolsthoorn) [inline tools/javascript/lib/commands/test.ts:68]:
```suggestion
    default: 0.9,
```

Related work uses 0.9 for this value

File: tools/javascript/lib/commands/test.ts [hunk 3]
```diff
@@ -32,13 +32,14 @@ export function getTestCommand(
 ): Command {
   const options = new Map<string, Yargs.Options>();
 
-  const commandGroup = "Type Inference Options:";
+  const typeInferenceGroup = "Type Inference Options:";
+  const variousProbabilitiesGroup = "Various Probabilities:";
```
Comment (mitchellolsthoorn) [inline tools/javascript/lib/commands/test.ts:36]: We can probably replace this by the sampling group introduced in  the constant pool PR #160

File: libraries/search-javascript/lib/testcase/statements/primitive/UndefinedStatement.ts [hunk 4]
```diff
@@ -26,29 +26,49 @@ import { JavaScriptTestCaseSampler } from "../../sampling/JavaScriptTestCaseSamp
  * @author Dimitri Stallenberg
  */
 export class UndefinedStatement extends PrimitiveStatement<boolean> {
```
Comment (mitchellolsthoorn) [inline libraries/search-javascript/lib/testcase/statements/primitive/UndefinedStatement.ts:28]: Undefined should probably not be a boolean

File: libraries/analysis-javascript/lib/target/TargetVisitor.ts (hunk 1 above)
Comment (dstallenberg) [reply libraries/analysis-javascript/lib/target/TargetVisitor.ts:407]: Done

File: libraries/search-javascript/lib/testcase/statements/primitive/UndefinedStatement.ts (hunk 4 above)
Comment (dstallenberg) [reply libraries/search-javascript/lib/testcase/statements/primitive/UndefinedStatement.ts]: Fixed.

**Review by mitchellolsthoorn — DISMISSED** [review]

**Review by mitchellolsthoorn — APPROVED** [review]

**Pull request merged**